src/app/pages/signup/signup.component.ts

```diff
@@ -1,67 +1,91 @@
-import { Component } from '@angular/core';
-import { FormBuilder, FormGroup, FormsModule, ReactiveFormsModule, Validators } from '@angular/forms';
-import { Router, RouterLink, RouterLinkActive } from '@angular/router';
+import { Component, OnInit, OnDestroy } from '@angular/core';
+import { AbstractControl, FormBuilder, FormGroup, Validators, ValidationErrors } from '@angular/forms';
+import { Router, RouterLink, RouterLinkActive, NavigationStart } from '@angular/router';
 import { AuthService } from '../../auth/auth.service';
-import { HttpClient, HttpHeaders, HttpErrorResponse } from '@angular/common/http';
-import { Observable, throwError, catchError, of } from 'rxjs';
+import { SharedModule } from '../../shared/shared.module';
+import { MatSnackBar } from '@angular/material/snack-bar';
+import { Subscription } from 'rxjs';
+import { defaultSnackbarConfig } from '../../shared/configs/defaultSnackbarConfig';
 
 @Component({
   selector: 'app-signup',
   standalone: true,
-  imports: [RouterLink, RouterLinkActive, FormsModule, ReactiveFormsModule],
+  imports: [RouterLink, RouterLinkActive, SharedModule],
   templateUrl: './signup.component.html',
   styleUrl: './signup.component.css'
 })
-export class SignupComponent {
+export class SignupComponent implements OnInit, OnDestroy {
   signupForm: FormGroup;
-  private apiUrl = 'https://rb-messenger.azurewebsites.net';
+  minUsernameLen = 4;
+  minPasswordLen = 6;
 
-  constructor(private fb: FormBuilder, private authService: AuthService,
-    private http: HttpClient, private router: Router
+  isInProgress: boolean = false;
+
+  private _routerSubscription!: Subscription;
+
+  constructor(private _fb: FormBuilder, private _authService: AuthService, private _router: Router,
+    private _snackBar: MatSnackBar
   ) {
-    this.signupForm = this.fb.group({
-      username: ['', Validators.required],
+    this.signupForm = this._fb.group({
+      username: ['', [Validators.required, Validators.minLength(this.minUsernameLen)]],
       email: ['', [Validators.required, Validators.email]],
-      password: ['', Validators.required],
-      repPassword: ['', Validators.required]
+      password: ['', [Validators.required, Validators.minLength(this.minPasswordLen), this.passwordValidator]],
+      confirmPassword: ['', [Validators.required]]
+    }, { validator: this.passwordMatchValidator });
+  }
+
+  ngOnInit() {
+    this._routerSubscription = this._router.events.subscribe(event => {
```

Comment:
Same here... For the future: If you see, that you copy the same code everywhere, then 90% you wrote smth wrong, or you should split the code to the reusable function/method.